[compiler-rt][ARM] Optimized f32 add/subtract for Armv6-M.

statham-arm opened the pull request:
This commit replaces the contents of the existing arm/addsf3.S with a much faster implementation that Arm has recently open-sourced in the Arm Optimized Routines git repository.

The new implementation is approximately 1.6× as fast as the old one on average. Some sample cycle timings from a Cortex-M0, with test cases covering both magnitude addition and subtraction and various cases of renormalization:

New code: 73,  63,  53,  81,  81
Old code: 83,  92,  88, 153, 168

This commit also contains a more thorough test suite for single precision addition and subtraction. Using that test suite I also found that the previous arm/addsf3.S had at least one bug, which the new code fixes: adding the largest denormal (0x007fffff) to itself returned 0x007ffffe, a slightly _smaller_ number, instead of the correct 0x00fffffe.

The test suite also includes thorough tests for the NaN handling policy implemented by the new code. This is in line with Arm's hardware FP implementations (so that switching between software and hardware FP makes as little difference as possible to the answers), but doesn't match what compiler-rt does in all other situations, so I've enabled it only under an `#ifdef` that should match when this implementation is selected.

The new code contains entry points for both addition and subtraction, with cross-branching between them after correcting signs. This avoids the overhead of treating subtraction as a sign-flipping wrapper on addition, but also means I had to add an extra piece of mechanism to the build scripts to allow the wrapper version of subsf3.c to be excluded from the build in the presence of the new addsf3.S. You can indicate that a platform-specific source file replaces an additional platform-independent one by setting its `crt_supersedes` property in cmake.

statham-arm committed:
[compiler-rt][ARM] Optimized f32 add/subtract for Armv6-M.

This commit replaces the contents of the existing arm/addsf3.S with a
much faster implementation that Arm has recently open-sourced in the
Arm Optimized Routines git repository.

The new implementation is approximately 1.6× as fast as the old one on
average. Some sample cycle timings from a Cortex-M0, with test cases
covering both magnitude addition and subtraction and various cases of
renormalization:

New code: 73,  63,  53,  81,  81
Old code: 83,  92,  88, 153, 168

This commit also contains a more thorough test suite for single
precision addition and subtraction. Using that test suite I also found
that the previous arm/addsf3.S had at least one bug, which the new
code fixes: adding the largest denormal (0x007fffff) to itself
returned 0x007ffffe, a slightly _smaller_ number, instead of the
correct 0x00fffffe.

The test suite also includes thorough tests for the NaN handling
policy implemented by the new code. This is in line with Arm's
hardware FP implementations (so that switching between software and
hardware FP makes as little difference as possible to the answers),
but doesn't match what compiler-rt does in all other situations, so
I've enabled it only under an `#ifdef` that should match when this
implementation is selected.

The new code contains entry points for both addition and subtraction,
with cross-branching between them after correcting signs. This avoids
the overhead of treating subtraction as a sign-flipping wrapper on
addition, but also means I had to add an extra piece of mechanism to
the build scripts to allow the wrapper version of subsf3.c to be
excluded from the build in the presence of the new addsf3.S. You can
indicate that a platform-specific source file replaces an additional
platform-independent one by setting its `crt_supersedes` property in
cmake.

statham-arm committed:
clang-format

smithp35 reviewed (commented):
A disclaimer that I was involved in the internal review for Arm Optimized Routines, and on principle I'm in favour of this change. I've concentrated on the parts unique to compiler-rt.

Do you happen to have a figure for code-size? One possible objection is someone preferring a smallest possible implementation for M0 at the expense of performance. Obviously can't just compare the length of the file as that contains comments, and instructions can be 2 or 4 bytes. To be clear I'm not advocating for an additional choice of routine based on size/performance.

Presumably the denormal issue you found was unique to the existing Arm assembly implementation and not the general C implementation (otherwise the tests would fail on non v6-m platforms)?

statham-arm committed:
Add comments in fnan2

statham-arm committed:
Rename __fnan2 to mention compiler-rt

statham-arm commented:
> Do you happen to have a figure for code-size? One possible objection is someone preferring a smallest possible implementation for M0 at the expense of performance.

You're right, the code size is bigger in this implementation. The new `addsf3.S` assembles to 648 bytes of code, and (at `-Os`) another 68 bytes for the helper `fnan2.c`. The old version was 312 bytes for `addsf3.S` and 22 bytes for the `subsf3.c` wrapper.

> Presumably the denormal issue you found was unique to the existing Arm assembly implementation and not the general C implementation (otherwise the tests would fail on non v6-m platforms)?

Yes, the C version in `lib/builtins/addsf3.c` (well, really `lib/builtins/fp_add_impl.inc`) passes the new test in full. (Or rather, correctly skips all the NaN test cases and passes the rest.)

statham-arm committed:
Lowercase instruction mnemonics

statham-arm committed:
Changes to fnan2 to be consistent wih #161546

statham-arm commented on compiler-rt/lib/builtins/arm/addsf3.S:
```diff
@@ -24,253 +24,829 @@
-  pop {r4, r5, r6, r7, pc}
-
-
+  PUSH {r4,r5,r6,lr}
```
OK, done.

statham-arm committed:
Conditionalize NaN-handling parts of tests

Now we only try to test the difficult parts if we're using the new
implementations, and otherwise, fall back to treating all NaNs as good
enough when a NaN result is expected.
